<launch>
  <!-- Arguments -->
  <arg name="slam_methods" default="rtab"/>
  <arg name="open_rviz" default="true"/>
  <arg name="localization"      default="false"/>
  <arg     if="$(arg localization)" name="rtabmap_args"  default=""/>
  <arg unless="$(arg localization)" name="rtabmap_args"  default="--delete_db_on_start"/>  

  <param name="use_sim_time" value="true"/>

  <!--P3dx World and Robot-->
  <include file="$(find p3dxbot)/launch/p3dxworld.launch" >
   </include>

   <node pkg="depthimage_to_laserscan" type="depthimage_to_laserscan" name="depthimage_to_laserscan" args="image:=/kinect_V2/depth/image_raw" >
    <param name="range_min" value="1.5"/>
    <param name="output_frame_id" value="p3dx/base_scan"/>
    <param name="scan_time" value = "0.005"/>
    <remap from="scan"        to="/kinect_scan"/>
    
  </node>

  <!-- SLAM -->
  <node pkg="tf" type="static_transform_publisher" name="kinect_broadcaster" args="0 0 0 0 0 0 1 p3dx/chassis p3dx/base_scan 10" />
  

    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="--delete_db_on_start">
          <param name="frame_id" type="string" value="p3dx/base_link"/>
          <param name="grid_size" type="double" value="50"/> <!-- 50 meters wide -->
        
          

          <param name="subscribe_depth" type="bool" value="true"/>
          <param name="subscribe_scan" type="bool" value="true"/>

          <remap from="odom" to="p3dx/odom"/>
          <remap from="scan" to="kinect_scan"/>

          <remap from="rgb/image"       to="/kinect_V2/rgb/image_raw"/>
          <remap from="depth/image"     to="/kinect_V2/depth/image_raw"/>
          <remap from="rgb/camera_info" to="/kinect_V2/rgb/camera_info"/>

          <param name="queue_size" type="int" value="10"/>

          <!-- RTAB-Map's parameters -->
          <param name="RGBD/AngularUpdate"        type="string" value="0.01"/>
          <param name="RGBD/LinearUpdate"         type="string" value="0.01"/>
          <param name="RGBD/OptimizeFromGraphEnd" type="string" value="false"/>
          
    </node>

      <!-- rviz -->
      <group if="$(arg open_rviz)"> 
        <node pkg="rviz" type="rviz" name="rviz" required="true"
              args="-d $(find p3dxbot)/rviz/p3dxbot_$(arg slam_methods).rviz"/>
      </group>
 

</launch>